### PR TITLE
[C#][VA/Skill] Integrate specific cultures

### PR DESCRIPTION
**Pull request by DiegoCardozo94**
**This Pull Request will be present in the next release.**

## Purpose

_What is the context of this pull request? Why is it being done?_

We updated the _Virtual Assistant_ and the _Skill_ language to replace the language and start to use complete cultures for each language with the objective of improve the functionality of the **multilanguage** implementation. 

## Changes
_Are there any changes that need to be called out as significant or particularly difficult to grasp? (Include illustrative screenshots for context if applicable.)_
- Implement `fallback to root language` functionality
- Rename different folders and files with each _specific culture_ name
- Use of the _specific culture_ name for the _deployment scripts_ files
- Changes in _mainDialog_ implementation to obtain the new locale language cultures (eg. 'en-us').
- Update template file to include the new changes when generate a new _VA_ and _Skill_ samples.

## Tests
_Is this covered by existing tests or new ones? If no, why not?_

It's covered by _LocalizationTests_, _MainDialogTests_ files.

## Testing steps

1. Deploy a _Virtual Assistant_ 
2. Verify in the bot solution that the resource folder contains the specific cultures
3. Execute the _Virtual Assistant_ 
4. Check the behavior of the _Virtual Assistant_ communication

## Feature Plan
_Are there any remaining steps or dependencies before this issue can be fully resolved? If so, describe and link to any relevant pull requests or issues._
*-*

#### General
- [ ] I have commented my code, particularly in hard-to-understand areas	
- [X] I have added or updated the appropriate tests	
- [ ] I have updated related documentation

#### Bots
- [X] I have replicated language model changes across the English, French, Italian, German, Spanish, and Chinese `.lu` files and validated that deployment is successful

#### Deployment Scripts
- [X] I have replicated my changes in the **Virtual Assistant Template** and **Sample** projects
- [X] I have replicated my changes in the **Skill Template** and **Sample** projects

### CODE REVIEW

**Comment by DiegoCardozo94** [comment]
We noticed that this PR contains conflicts, we'll update fixing the merge conflicts

**Comment by DiegoCardozo94** [comment]
Hi team, this Pull Request is ready for review.

**Comment by lauren-mills** [comment]
@DiegoCardozo94 does this work correctly if you specify a different culture in the emulator, en-gb or es-mx for example?

**Comment by DiegoCardozo94** [comment]
Hi @lauren-mills!

Following the [Supported Languages of LUIS](https://docs.microsoft.com/en-us/azure/cognitive-services/luis/luis-language-support#languages-supported), `en-gb` is not a specifically a region with utterances which `LUIS` understands.
On the other hand, `es-mx` is a supported one by `LUIS` and we should add a new `locale` folder with the intents and utterances related to that culture. After that, the specific culture `es-mx` will work correctly.

Does it correct for you?

**Comment by lauren-mills** [comment]
@DiegoCardozo94 I was referring to the ability to fallback to a top level language, regardless of what the user passes as their full locale. For instance, if a user's locale is set to en-gb by their client, we should be able to use the en-us language model and responses because they have the same root language, rather than saying its unsupported. Likewise with es-es and es-mx.

Otherwise, we will not have flexibility in supporting the various locales we might be sent by a client.

**Comment by DiegoCardozo94** [comment]
@lauren-mills, good catch! We will be analyzing the scenario that you are mentioning and get back to you later.

**Comment by dfavretto** [comment]
@lauren-mills we added a task to our backlog to implement the fallback functionality, and we'll let you know when this is done so you can give it a try. Thanks for the feedback :)

**Comment by dfavretto** [comment]
Hi @lauren-mills, we finished to implement the **fallback to root language**, please take a look at it. 

Just to be sure, these are the steps we used for testing with `en-gb`:
1. Deploy a Virtual Assistant with `en-us`.
2. Change the configuration in the emulator to a non-existing locale, in this case `en-gb`.
3. Talk to the bot and check if the activity sent have the locale `en-gb`.
image
4. The bot will fallback to `en` root language and answer correctly.
	
Our next step would be to add the necessary tests to cover this feature.

File: templates/Virtual-Assistant-Template/csharp/Sample/VirtualAssistantSample/Dialogs/MainDialog.cs [hunk 1]
```diff
@@ -160,6 +159,12 @@ protected override async Task<InterruptionAction> OnInterruptDialogAsync(DialogC
                             }
                     }
                 }
+                catch (Exception)
+                {
+                    var template = _templateEngine.EvaluateTemplate("confusedMessage");
+                    var response = await _activityGenerator.CreateActivityFromText(template, null, dc.Context, _langGenerator);
+                    await dc.Context.SendActivityAsync(response);
+                }
```
Comment (DiegoCardozo94) [inline templates/Virtual-Assistant-Template/csharp/Sample/VirtualAssistantSample/Dialogs/MainDialog.cs]: @lauren-mills, we added this for handling the error thrown when the selected locale doesn't exist

File: templates/Virtual-Assistant-Template/csharp/Sample/VirtualAssistantSample/Dialogs/MainDialog.cs [hunk 2]
```diff
@@ -194,30 +199,36 @@ protected override async Task OnMessageActivityAsync(DialogContext innerDc, Canc
...
+                    else
+                    {
+                        throw new Exception();
+                    }
                 }
-                else
+                catch (Exception)
                 {
                     var template = _templateEngine.EvaluateTemplate("confusedMessage");
                     var response = await _activityGenerator.CreateActivityFromText(template, null, innerDc.Context, _langGenerator);
```
Comment (DiegoCardozo94) [inline templates/Virtual-Assistant-Template/csharp/Sample/VirtualAssistantSample/Dialogs/MainDialog.cs]: @lauren-mills, we added this for handling the error thrown when the selected locale doesn't exist

File: templates/Virtual-Assistant-Template/csharp/Sample/VirtualAssistantSample/Dialogs/MainDialog.cs (hunk 1 above)
Comment (DiegoCardozo94) [inline templates/Virtual-Assistant-Template/csharp/Sample/VirtualAssistantSample/Dialogs/MainDialog.cs]: We added this for handling the error thrown when the selected locale doesn't exist.

File: templates/Virtual-Assistant-Template/csharp/Sample/VirtualAssistantSample/Dialogs/MainDialog.cs (hunk 2 above)
Comment (DiegoCardozo94) [inline templates/Virtual-Assistant-Template/csharp/Sample/VirtualAssistantSample/Dialogs/MainDialog.cs]: We added this for handling the error thrown when the selected locale doesn't exist.

**Review by DiegoCardozo94 — COMMENTED** [review]
Hi @lauren-mills, we did some changes out of the scope of **specific cultures** that were needed for correctly handling the error and implement the tests. Please take a look at these changes and let us know if it looks fine for you.

Just for clarification, as the GitHub diff seems a bit messy, we only encapsulated the `OnInterruptDialogAsync` and `OnMessageActivityAsync` methods with `try`-`catch`, but the change in indentation messes up the diff view.

File: templates/Skill-Template/csharp/Sample/SkillSample/Dialogs/MainDialog.cs [hunk 3]
```diff
@@ -241,5 +238,19 @@ private async Task PopulateStateFromSemanticAction(ITurnContext context)
             //    state.CurrentCoordinates = locationObj;
             // }
         }
+
```
Comment (lauren-mills) [inline templates/Skill-Template/csharp/Sample/SkillSample/Dialogs/MainDialog.cs]: Might be good to move this method into the BotServices class so all the logic to get the cognitive models is in one place.

**Review by lauren-mills — APPROVED** [review]

File: templates/Skill-Template/csharp/Sample/SkillSample/Dialogs/MainDialog.cs (hunk 3 above)
Comment (DiegoCardozo94) [reply templates/Skill-Template/csharp/Sample/SkillSample/Dialogs/MainDialog.cs]: Sure, we'll make these changes in both pull request.

**Review by lzc850612 — APPROVED** [review]

**Comment by Batta32** [comment]
Hi team, we already solved the merge conflicts.

**Comment by Batta32** [comment]
Hi team, we identified new merge conflicts that we already **solved**.

**Comment by lauren-mills** [comment]
Seems like there are some issues in the tests after the build. Can you take a look? We changed the response localization in the templates to use LG, so the files have changed from when this PR started.

**Comment by Batta32** [comment]
Thanks @lauren-mills, we will validate the test execution and get back to you later 😊.

**Comment by DiegoCardozo94** [comment]
Hi @lauren-mills,

We **fixed the tests** in Virtual Assistant and Skill Samples. However, we **didn't replicate the changes** in the templates because we identified **inconsistences** (e.g. `.lg` files are not present in the template). 

If you consider necessary to replicate the functionality of specific cultures in the templates, please let us know. 😊

**Pull request merged**